### DISCUSSION TITLE
How can I display marker labels?

### DISCUSSION THREAD

**Discussion by zdefne-usgs** [TerriaJS/terriajs#6118] [Q&A]
How can I display only labels for a marker? Like having a CSV input for coordinates of a city and displaying city name on top its marker?

**Comment by nf-s** [comment]
Hi @zdefne-usgs 

Currently we don't support labels for CSV.

There are two other options:
- CZML file
- GeoJson with CZML tempalte

## CZML file

TerriaJS supports CZML which can be used to draw points and labels.

So you can generate your own CZML files

See https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/CZML-Guide for file structure

Example - http://ci.terria.io/main/#clean&https://gist.githubusercontent.com/nf-s/4083f4855cabc102861986576167b835/raw/f06669e6475b6238994e8407dd8cd4ce9713d852/czml-point-label.json&share=s-1xJ9azjUAuBkfkTblnzw43fW4ic

### Example CZML

```json
[
  {
    "id": "document",
    "name": "CZML Point",
    "version": "1.0"
  },
  {
    "id": "point 1",
    "name": "point",
    "position": {
      "cartographicDegrees": [-111.0, 40.0, 0]
    },
    "point": {
      "color": {
        "rgba": [255, 255, 255, 255]
      },
      "outlineColor": {
        "rgba": [255, 0, 0, 255]
      },
      "outlineWidth": 4,
      "pixelSize": 20
    },
    "label": {
      "style": {
        "labelStyle": "FILL_AND_OUTLINE"
      },
      "fillColor": {
        "rgba": [0, 0, 0, 255]
      },
      "outlineColor": {
        "rgba": [255, 255, 255, 255]
      },
      "outlineWidth": 2,
      "text": "Hello",
      "pixelOffset": {
        "cartesian2": [60, 0]
      }
    }
  }
]
```

## GeoJSON with CZML Template 

Alternatively, we have `czmlTemplate` in `GeoJsonCatalogItem` - this can be used to add points/labels to a geoJSON file

Example - http://ci.terria.io/main/#clean&https://gist.githubusercontent.com/nf-s/182d099fd57cfd7c22d9707741048f9e/raw/4a2e3a480bcb2310c10445d926687b90f7a7ffde/geojson-czml-label.json&share=s-rpipt83S2ME8vbozlPhe6GHEL6r

### Catalog JSON example

`"czmlTemplate"` will create a point (with 10 pixels diameter) with label showing the value of `"someOtherProp"` in each point feature

```json
{
      "id": "geojson-czml-label",
      "name": "geojson-czml-label",
      "url": "build/TerriaJS/test/GeoJSON/points.geojson",
      "type": "geojson",
      "czmlTemplate": {
        "point": {
          "pixelSize": 10
        },
        "label": {
          "horizontalOrigin": "LEFT",
          "heightReference": {
            "heightReference": "RELATIVE_TO_GROUND"
          },
          "scale": 1,
          "style": {
            "labelStyle": "FILL_AND_OUTLINE"
          },
          "fillColor": {
            "rgba": [
              0,
              0,
              0,
              255
            ]
          },
          "outlineColor": {
            "rgba": [
              255,
              255,
              255,
              255
            ]
          },
          "outlineWidth": 2,
          "text": {
            "reference": "#properties.someOtherProp"
          },
          "eyeOffset": {
            "cartesian": [
              10,
              10,
              0
            ]
          }
        }
      }
    }
```

### GeoJSON file example

```json
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {
        "height": 10,
        "radius": 10,
        "someOtherProp": "what"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [
          144.91734981536865,
          -37.824700770115996
        ]
      }
    },
...
```

**Comment by pjanik** [comment]
@nf-s, your examples work well in 3D view. However, after switching to 2D, labels become very small, blurry, and generally not readable. They don't seem to follow the spec from CZML. Is it a known issue? Is there any workaround (eg separate styles for 2d view)?
